# Football Prediction Model

Creating a model to predict football outcomes, such as correct scores, match results (win, lose, or draw), and head-to-head outcomes using data from **Football-Data.co.ke** or similar sources requires several steps. Here's a detailed guide:

### 1. **Data Collection and Preprocessing**
First, gather the required data from **Football-Data.co.ke**. You need match data including:
- Team names
- Dates of matches
- Match outcomes (home win, away win, draw)
- Goals scored by each team (correct score)
- Head-to-head match results

This data can often be found in downloadable CSV files on the site. You can then load it into Python using `Pandas`.

### Import necessary libraries

In [5]:
import pandas as pd
import os
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt 


## Merge the excell files

In [6]:

# get list of excell files
folder_path = '/home/sirian/Documents/AI_Agents_&_LLMS/Cursor/sirian/data'
files = [f for f in os.listdir(folder_path) if f.endswith('.xls') or f.endswith('.xlsx') ]

# load data to dataframes
dataframes = []
for file in files:
    df = pd.read_excel(os.path.join(folder_path, file))
    dataframes.append(df)
    
# merge dataframes    
merged_df = pd.concat(dataframes, ignore_index=True)

#export dataframe to csv
output_path = '/home/sirian/Documents/AI_Agents_&_LLMS/Cursor/sirian/data'
merged_df.to_csv(os.path.join(output_path, 'merged_data_1.csv'), index=False)

print('Data merged and saved as merged_data_1.csv')

Data merged and saved as merged_data_1.csv


column names explained
HG - Home Goals
AG - Away goals
Res - Result
PSCH - 

## Notes for Football Data

All data is in csv format, ready for use within standard spreadsheet applications. Please note that some abbreviations are no longer in use (in particular odds from specific bookmakers no longer used) and refer to data collected in earlier seasons. For a current list of what bookmakers are included in the dataset please visit http://www.football-data.co.uk/matches.php

Key to results data:

- Div = League Division
- Date = Match Date (dd/mm/yy)
- Time = Time of match kick off
- HomeTeam = Home Team
- AwayTeam = Away Team
- FTHG and HG = Full Time Home Team Goals
- FTAG and AG = Full Time Away Team Goals
- FTR and Res = Full Time Result (H=Home Win, D=Draw, A=Away Win)
- HTHG = Half Time Home Team Goals
- HTAG = Half Time Away Team Goals
- HTR = Half Time Result (H=Home Win, D=Draw, A=Away Win)

- Match Statistics (where available)
- Attendance = Crowd Attendance
- Referee = Match Referee
- HS = Home Team Shots
- AS = Away Team Shots
- HST = Home Team Shots on Target
- AST = Away Team Shots on Target
- HHW = Home Team Hit Woodwork
- AHW = Away Team Hit Woodwork
- HC = Home Team Corners
- AC = Away Team Corners
- HF = Home Team Fouls Committed
- AF = Away Team Fouls Committed
- HFKC = Home Team Free Kicks Conceded
- AFKC = Away Team Free Kicks Conceded
- HO = Home Team Offsides
- AO = Away Team Offsides
- HY = Home Team Yellow Cards
- AY = Away Team Yellow Cards
- HR = Home Team Red Cards
- AR = Away Team Red Cards
- HBP = Home Team Bookings Points (10 = yellow, 25 = red)
- ABP = Away Team Bookings Points (10 = yellow, 25 = red)

Note that Free Kicks Conceeded includes fouls, offsides and any other offense commmitted and will always be equal to or higher than the number of fouls. Fouls make up the vast majority of Free Kicks Conceded. Free Kicks Conceded are shown when specific data on Fouls are not available (France 2nd, Belgium 1st and Greece 1st divisions).

Note also that English and Scottish yellow cards do not include the initial yellow card when a second is shown to a player converting it into a red, but this is included as a yellow (plus red) for European games.


[Pre-closing odds. For closing odds, see further below.]

Key to 1X2 (match) betting odds data: 

- B365H = Bet365 home win odds
- B365D = Bet365 draw odds
- B365A = Bet365 away win odds
- BFH = Betfair home win odds
- BFD = Betfair draw odds
- BFA = Betfair away win odds
- BSH = Blue Square home win odds
- BSD = Blue Square draw odds
- BSA = Blue Square away win odds
- BWH = Bet&Win home win odds
- BWD = Bet&Win draw odds
- BWA = Bet&Win away win odds
- GBH = Gamebookers home win odds
- GBD = Gamebookers draw odds
- GBA = Gamebookers away win odds
- IWH = Interwetten home win odds
- IWD = Interwetten draw odds
- IWA = Interwetten away win odds
- LBH = Ladbrokes home win odds
- LBD = Ladbrokes draw odds
- LBA = Ladbrokes away win odds
- PSH and PH = Pinnacle home win odds
- PSD and PD = Pinnacle draw odds
- PSA and PA = Pinnacle away win odds
- SOH = Sporting Odds home win odds
- SOD = Sporting Odds draw odds
- SOA = Sporting Odds away win odds
- SBH = Sportingbet home win odds
- SBD = Sportingbet draw odds
- SBA = Sportingbet away win odds
- SJH = Stan James home win odds
- SJD = Stan James draw odds
- SJA = Stan James away win odds
- SYH = Stanleybet home win odds
- SYD = Stanleybet draw odds
- SYA = Stanleybet away win odds
- VCH = VC Bet home win odds
- VCD = VC Bet draw odds
- VCA = VC Bet away win odds
- WHH = William Hill home win odds
- WHD = William Hill draw odds
- WHA = William Hill away win odds

### Bb1X2 = Number of BetBrain bookmakers used to calculate match odds averages and maximums
- BbMxH = Betbrain maximum home win odds
- BbAvH = Betbrain average home win odds
- BbMxD = Betbrain maximum draw odds
- BbAvD = Betbrain average draw win odds
- BbMxA = Betbrain maximum away win odds
- BbAvA = Betbrain average away win odds

- MaxH = Market maximum home win odds
- MaxD = Market maximum draw win odds
- MaxA = Market maximum away win odds
- AvgH = Market average home win odds
- AvgD = Market average draw win odds
- AvgA = Market average away win odds

- BFEH = Betfair Exchange home win odds
- BFED = Betfair Exchange draw odds
- BFEA = Betfair Exchange away win odds



### Key to total goals betting odds:

- BbOU = Number of BetBrain bookmakers used to calculate over/under 2.5 goals (total goals) averages and maximums
- BbMx>2.5 = Betbrain maximum over 2.5 goals
- BbAv>2.5 = Betbrain average over 2.5 goals
- BbMx<2.5 = Betbrain maximum under 2.5 goals 
- BbAv<2.5 = Betbrain average under 2.5 goals

- GB>2.5 = Gamebookers over 2.5 goals
- GB<2.5 = Gamebookers under 2.5 goals
- B365>2.5 = Bet365 over 2.5 goals
- B365<2.5 = Bet365 under 2.5 goals
- P>2.5 = Pinnacle over 2.5 goals
- P<2.5 = Pinnacle under 2.5 goals
- Max>2.5 = Market maximum over 2.5 goals
- Max<2.5 = Market maximum under 2.5 goals
- Avg>2.5 = Market average over 2.5 goals
- Avg<2.5 = Market average under 2.5 goals



Key to Asian handicap betting odds:

- BbAH = Number of BetBrain bookmakers used to Asian handicap averages and maximums
- BbAHh = Betbrain size of handicap (home team)
- AHh = Market size of handicap (home team) (since 2019/2020)
- BbMxAHH = Betbrain maximum Asian handicap home team odds
- BbAvAHH = Betbrain average Asian handicap home team odds
- BbMxAHA = Betbrain maximum Asian handicap away team odds
- BbAvAHA = Betbrain average Asian handicap away team odds

- GBAHH = Gamebookers Asian handicap home team odds
- GBAHA = Gamebookers Asian handicap away team odds
- GBAH = Gamebookers size of handicap (home team)
- LBAHH = Ladbrokes Asian handicap home team odds
- LBAHA = Ladbrokes Asian handicap away team odds
- LBAH = Ladbrokes size of handicap (home team)
- B365AHH = Bet365 Asian handicap home team odds
- B365AHA = Bet365 Asian handicap away team odds
- B365AH = Bet365 size of handicap (home team)
- PAHH = Pinnacle Asian handicap home team odds
- PAHA = Pinnacle Asian handicap away team odds
- MaxAHH = Market maximum Asian handicap home team odds
- MaxAHA = Market maximum Asian handicap away team odds	
- AvgAHH = Market average Asian handicap home team odds
- AvgAHA = Market average Asian handicap away team odds



Closing odds (last odds before match starts)

As above but with an additional "C" character following the bookmaker abbreviation/Max/Avg

Football-Data would like to acknowledge the following sources which have been utilised in the compilation of Football-Data's results and odds files.


Current results (full time, half time)
XScores - http://www.xscores .com

Match statistics
BBC, ESPN Soccer, Bundesliga.de, Gazzetta.it and Football.fr

### Bookmakers betting odds
Betbrain.com
Oddsportal.com
Individual bookmakers

Betting odds for weekend games are collected Friday afternoons, and on Tuesday afternoons for midweek games.

Additional match statistics (corners, shots, bookings, referee etc.) for the 2000/01 and 2001/02 seasons for the English, Scottish and German leagues were provided by Sports.com (now under new ownership and no longer available).


### Created a separate odds dataframe separated by date

In [10]:
# Read the merged data with low_memory=False to handle mixed types
print("Reading merged_data_1.csv...")
df = pd.read_csv('data/merged_data_1.csv', low_memory=False)

print(f"Original data shape: {df.shape}")

# Convert Date column to datetime, handling any invalid dates
df['Date'] = pd.to_datetime(df['Date'], errors='coerce')

# Remove rows with invalid dates
df = df.dropna(subset=['Date'])

print(f"Data shape after cleaning dates: {df.shape}")
print(f"Date range: {df['Date'].min()} to {df['Date'].max()}")

# Define odds-related columns based on the documentation
odds_columns = [
    # Basic match odds (1X2 betting)
    'GBH', 'GBD', 'GBA', 'IWH', 'IWD', 'IWA', 'LBH', 'LBD', 'LBA', 
    'SBH', 'SBD', 'SBA', 'WHH', 'WHD', 'WHA', 'SYH', 'SYD', 'SYA',
    'B365H', 'B365D', 'B365A', 'SOH', 'SOD', 'SOA', 'BWH', 'BWD', 'BWA',
    'SJH', 'SJD', 'SJA', 'VCH', 'VCD', 'VCA', 'BSH', 'BSD', 'BSA',
    'PSH', 'PSD', 'PSA', 'PSCH', 'PSCD', 'PSCA',
    
    # Total goals betting odds
    'GB>2.5', 'GB<2.5', 'B365>2.5', 'B365<2.5', 'P>2.5', 'P<2.5',
    'PC>2.5', 'PC<2.5', 'B365C>2.5', 'B365C<2.5',
    
    # Asian handicap betting odds
    'GBAHH', 'GBAHA', 'GBAH', 'LBAHH', 'LBAHA', 'LBAH', 'B365AHH', 'B365AHA', 'B365AH',
    'PAHH', 'PAHA', 'B365CAHH', 'B365CAHA', 'PCAHH', 'PCAHA',
    
    # BetBrain aggregated odds
    'Bb1X2', 'BbMxH', 'BbAvH', 'BbMxD', 'BbAvD', 'BbMxA', 'BbAvA',
    'BbOU', 'BbMx>2.5', 'BbAv>2.5', 'BbMx<2.5', 'BbAv<2.5',
    'BbAH', 'BbAHh', 'BbMxAHH', 'BbAvAHH', 'BbMxAHA', 'BbAvAHA',
    
    # Market aggregated odds
    'MaxH', 'MaxD', 'MaxA', 'AvgH', 'AvgD', 'AvgA',
    'Max>2.5', 'Max<2.5', 'Avg>2.5', 'Avg<2.5',
    'MaxAHH', 'MaxAHA', 'AvgAHH', 'AvgAHA',
    'MaxCH', 'MaxCD', 'MaxCA', 'AvgCH', 'AvgCD', 'AvgCA',
    'MaxC>2.5', 'MaxC<2.5', 'AvgC>2.5', 'AvgC<2.5',
    'MaxCAHH', 'MaxCAHA', 'AvgCAHH', 'AvgCAHA',
    
    # Closing odds (with 'C' suffix)
    'B365CH', 'B365CD', 'B365CA', 'BWCH', 'BWCD', 'BWCA',
    'IWCH', 'IWCD', 'IWCA', 'WHCH', 'WHCD', 'WHCA',
    'VCCH', 'VCCD', 'VCCA', 'AHCh',
    
    # Betfair odds
    'BFH', 'BFD', 'BFA', '1XBH', '1XBD', '1XBA', 'BFEH', 'BFED', 'BFEA',
    'BFE>2.5', 'BFE<2.5', 'BFEAHH', 'BFEAHA', 'BFCH', 'BFCD', 'BFCA',
    '1XBCH', '1XBCD', '1XBCA', 'BFECH', 'BFECD', 'BFECA',
    'BFEC>2.5', 'BFEC<2.5', 'BFECAHH', 'BFECAHA'
]

# Filter columns that actually exist in the dataset
existing_odds_columns = [col for col in odds_columns if col in df.columns]
print(f"Found {len(existing_odds_columns)} odds-related columns out of {len(odds_columns)} defined")

# Create odds dataframe with Date and all odds columns
odds_data = df[['Date'] + existing_odds_columns].copy()

# Remove rows where all odds columns are NaN (no odds data available)
odds_data_clean = odds_data.dropna(subset=existing_odds_columns, how='all')

print(f"Odds data shape after removing rows with no odds: {odds_data_clean.shape}")

# Save the odds data
odds_data_clean.to_csv('data/odds_data.csv', index=False)
print("Odds data saved to data/odds_data.csv")

# Show some statistics
print(f"\nOdds data statistics:")
print(f"Date range: {odds_data_clean['Date'].min()} to {odds_data_clean['Date'].max()}")
print(f"Number of unique dates: {odds_data_clean['Date'].nunique()}")

# Show which odds columns have the most data
odds_coverage = odds_data_clean[existing_odds_columns].notna().sum().sort_values(ascending=False)
print(f"\nTop 10 odds columns by data coverage:")
print(odds_coverage.head(10)) 

Reading merged_data_1.csv...
Original data shape: (9342, 187)
Data shape after cleaning dates: (9170, 187)
Date range: 1999-08-07 00:00:00 to 2024-09-22 00:00:00
Found 155 odds-related columns out of 155 defined
Odds data shape after removing rows with no odds: (8790, 156)
Odds data saved to data/odds_data.csv

Odds data statistics:
Date range: 2000-08-19 00:00:00 to 2024-09-22 00:00:00
Number of unique dates: 2438

Top 10 odds columns by data coverage:
WHH      8676
WHD      8676
WHA      8676
IWH      8541
IWD      8541
IWA      8541
B365A    8030
B365D    8030
B365H    8030
BWH      7267
dtype: int64


- odds data is now separate from match data.
- Odds is a predictive feature.
- The assumption is that these are pregame odds.
- pregame odds are show likely hood a team is either going to win, draw or lose

### match data saved to `match_statistics.csv`


In [9]:
# Read the merged data
print("Reading merged_data_1.csv...")
df = pd.read_csv('data/merged_data_1.csv', low_memory=False)

print(f"Original data shape: {df.shape}")

# Convert Date column to datetime, handling any invalid dates
df['Date'] = pd.to_datetime(df['Date'], errors='coerce')

# Remove rows with invalid dates
df = df.dropna(subset=['Date'])

print(f"Data shape after cleaning dates: {df.shape}")

# Define odds-related columns to remove (same as in separate_odds_data.py)
odds_columns_to_remove = [
    # Basic match odds (1X2 betting)
    'GBH', 'GBD', 'GBA', 'IWH', 'IWD', 'IWA', 'LBH', 'LBD', 'LBA', 
    'SBH', 'SBD', 'SBA', 'WHH', 'WHD', 'WHA', 'SYH', 'SYD', 'SYA',
    'B365H', 'B365D', 'B365A', 'SOH', 'SOD', 'SOA', 'BWH', 'BWD', 'BWA',
    'SJH', 'SJD', 'SJA', 'VCH', 'VCD', 'VCA', 'BSH', 'BSD', 'BSA',
    'PSH', 'PSD', 'PSA', 'PSCH', 'PSCD', 'PSCA',
    
    # Total goals betting odds
    'GB>2.5', 'GB<2.5', 'B365>2.5', 'B365<2.5', 'P>2.5', 'P<2.5',
    'PC>2.5', 'PC<2.5', 'B365C>2.5', 'B365C<2.5',
    
    # Asian handicap betting odds
    'GBAHH', 'GBAHA', 'GBAH', 'LBAHH', 'LBAHA', 'LBAH', 'B365AHH', 'B365AHA', 'B365AH',
    'PAHH', 'PAHA', 'B365CAHH', 'B365CAHA', 'PCAHH', 'PCAHA',
    
    # BetBrain aggregated odds
    'Bb1X2', 'BbMxH', 'BbAvH', 'BbMxD', 'BbAvD', 'BbMxA', 'BbAvA',
    'BbOU', 'BbMx>2.5', 'BbAv>2.5', 'BbMx<2.5', 'BbAv<2.5',
    'BbAH', 'BbAHh', 'BbMxAHH', 'BbAvAHH', 'BbMxAHA', 'BbAvAHA',
    
    # Market aggregated odds
    'MaxH', 'MaxD', 'MaxA', 'AvgH', 'AvgD', 'AvgA',
    'Max>2.5', 'Max<2.5', 'Avg>2.5', 'Avg<2.5',
    'MaxAHH', 'MaxAHA', 'AvgAHH', 'AvgAHA',
    'MaxCH', 'MaxCD', 'MaxCA', 'AvgCH', 'AvgCD', 'AvgCA',
    'MaxC>2.5', 'MaxC<2.5', 'AvgC>2.5', 'AvgC<2.5',
    'MaxCAHH', 'MaxCAHA', 'AvgCAHH', 'AvgCAHA',
    
    # Closing odds (with 'C' suffix)
    'B365CH', 'B365CD', 'B365CA', 'BWCH', 'BWCD', 'BWCA',
    'IWCH', 'IWCD', 'IWCA', 'WHCH', 'WHCD', 'WHCA',
    'VCCH', 'VCCD', 'VCCA', 'AHCh',
    
    # Betfair odds
    'BFH', 'BFD', 'BFA', '1XBH', '1XBD', '1XBA', 'BFEH', 'BFED', 'BFEA',
    'BFE>2.5', 'BFE<2.5', 'BFEAHH', 'BFEAHA', 'BFCH', 'BFCD', 'BFCA',
    '1XBCH', '1XBCD', '1XBCA', 'BFECH', 'BFECD', 'BFECA',
    'BFEC>2.5', 'BFEC<2.5', 'BFECAHH', 'BFECAHA'
]

# Filter columns that actually exist in the dataset
existing_odds_columns = [col for col in odds_columns_to_remove if col in df.columns]
print(f"Found {len(existing_odds_columns)} odds-related columns to remove out of {len(odds_columns_to_remove)} defined")

# Remove odds columns to keep only match statistics
match_stats_df = df.drop(columns=existing_odds_columns)

print(f"Match statistics data shape after removing odds columns: {match_stats_df.shape}")

# Show remaining columns
print(f"\nRemaining columns ({len(match_stats_df.columns)}):")
print(list(match_stats_df.columns))

# Save the cleaned match statistics data
match_stats_df.to_csv('data/match_statistics.csv', index=False)
print("\nMatch statistics data saved to data/match_statistics.csv")

# Show some statistics about the cleaned data
print(f"\nMatch statistics data summary:")
print(f"Date range: {match_stats_df['Date'].min()} to {match_stats_df['Date'].max()}")
print(f"Number of unique dates: {match_stats_df['Date'].nunique()}")
print(f"Number of unique teams (HomeTeam): {match_stats_df['HomeTeam'].nunique()}")
print(f"Number of unique teams (AwayTeam): {match_stats_df['AwayTeam'].nunique()}")

# Show data types of remaining columns
print(f"\nData types of remaining columns:")
print(match_stats_df.dtypes.value_counts())

# Show a sample of the cleaned data
print(f"\nSample of cleaned match statistics data:")
print(match_stats_df.head()) 

Reading merged_data_1.csv...
Original data shape: (9342, 187)
Data shape after cleaning dates: (9170, 187)
Found 155 odds-related columns to remove out of 155 defined
Match statistics data shape after removing odds columns: (9170, 32)

Remaining columns (32):
['Div', 'Date', 'HomeTeam', 'AwayTeam', 'FTHG', 'FTAG', 'FTR', 'HTHG', 'HTAG', 'HTR', 'Attendance', 'Referee', 'HS', 'AS', 'HST', 'AST', 'HHW', 'AHW', 'HC', 'AC', 'HF', 'AF', 'HO', 'AO', 'HY', 'AY', 'HR', 'AR', 'HBP', 'ABP', 'Time', 'AHh']

Match statistics data saved to data/match_statistics.csv

Match statistics data summary:
Date range: 1999-08-07 00:00:00 to 2024-09-22 00:00:00
Number of unique dates: 2542
Number of unique teams (HomeTeam): 48
Number of unique teams (AwayTeam): 48

Data types of remaining columns:
float64           24
object             7
datetime64[ns]     1
Name: count, dtype: int64

Sample of cleaned match statistics data:
  Div       Date       HomeTeam     AwayTeam  FTHG  FTAG FTR  HTHG  HTAG HTR  \
0  E0

 - remained with purely match data separate from odds data

**Summary**

I've successfully:
1. **Separated the odds data** from merged_data_1.csv and saved it as data/odds_data.csv (8,790 rows with 156 columns including Date and all odds-related columns)

2. **Created a cleaned match statistics dataset** by removing all odds columns from merged_data_1.csv and saved it as data/match_statistics.csv (9,170 rows with 32 columns)

The remaining columns in the match statistics dataset are:
- **Basic match info**: Div, Date, HomeTeam, AwayTeam, Time
- **Goals**: FTHG, FTAG, FTR, HTHG, HTAG, HTR
- **Match statistics**: Attendance, Referee, HS, AS, HST, AST, HHW, AHW, HC, AC, HF, AF, HO, AO, HY, AY, HR, AR, HBP, ABP, AHh

You can now copy the code above into a new cell in your main.ipynb notebook to run the data cleaning process. The code will remove all odds-related columns and keep only the pure match statistics data